File: 25.12.25/3sum_closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int sum = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length; i++) {
            int l = i + 1;
            int r = nums.length - 1;
            while (l < r) {
                int currentSum = nums[i] + nums[l] + nums[r];
                if (currentSum > target) {
                    r--;
                }else if(currentSum < target){
                    l++;
                }else{
                    return currentSum;
                }
                sum = Math.abs(currentSum - target) < Math.abs(sum - target) ? currentSum : sum;
            }
        }
        return sum;
    }
}

File: 29.12.25/3sum_closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        if len(nums) < 3:
            raise ValueError("Need at least 3 numbers")

        nums.sort()
        closest = float('inf')

        for i in range(len(nums) - 2):          # stop at len-3
            left, right = i + 1, len(nums) - 1

            while left < right:
                curr_sum = nums[i] + nums[left] + nums[right]

                # update closest if this sum is nearer to target
                if abs(curr_sum - target) < abs(closest - target):
                    closest = curr_sum

                if curr_sum < target:
                    left += 1
                elif curr_sum > target:
                    right -= 1
                else:
                    # exact match
                    return curr_sum

        return closest
